### ISSUE TITLE
Feature Proposal: Add directory level all targets

### ISSUE THREAD

**Issue by jml1795** [mesonbuild/meson#4977]
It would be helpful to have a directory level all target added automatically (for ninja backend) to allow rebuilds of all packages in that directory and subdirectories.

For example, if a directory contained 3 shared libraries and two sub-directories, invoking `ninja some_dir/all` would have the three libraries and each subdirectory all target as prerequisites. This would also simplify rebuilding a single target in a folder by running `ninja some/path/all` instead of `ninja some/path/libsome_really_long_name.so`

This may already be a feature and I have missed it; if so, please let me know.

I have several questions about how to implement this:
1) Can this be added by default instead of having to opt in at each folder?
1.a) If not added by default, maybe a project option (something like `package_targets=true` in the project declaration) so we don't have to explicitly provide a list.
2) Is this applicable to MSVC (I don't have a windows box to test the implementation on)?
3) Is there another name we would prefer? For example `some_dir/**` or `some_dir/@@` (`**` would need to be escaped or quoted in most shells)

**Comment by obilaniu** [comment]
Can you make use of
- [`run_target()`](https://mesonbuild.com/Reference-manual.html#run_target) with your choice of `depends:` and a dummy `command:`, or
- [`custom_target()`](https://mesonbuild.com/Reference-manual.html#custom_target)?

Maybe even both?

**Comment by jml1795** [comment]
possibly, but that would require having two lists of essentially the same data if I am understanding correctly. I.e. (my best pseudo-code)
```meson
subdir('a')
subdir('b')
...
subdir('z')

all_tgt = run_target('all', dummy_cmd, depends: [
  a_dep,
  b_dep,
  ...
  z_dep,
])
```
Is this what you mean? Aside from having to list everything twice, it also leaves open the possibility of getting the packages out of sync. Also not sure if you would be able to have an `all` target in two different directories or if you would need a `dir1_all` and `dir2_all`

**Comment by obilaniu** [comment]
That is what I had in mind, but if that's not good enough then I don't know what to suggest. Maybe you could add a `run_target()` that would be a python script that invokes [`meson introspect`](https://mesonbuild.com/IDE-integration.html), then parses the information in the JSON therein to decide which ninja commands to execute according to your wishes.

**Comment by jpakkane** [comment]
> This would also simplify rebuilding a single target in a folder by running `ninja some/path/all` instead of `ninja some/path/libsome_really_long_name.so`

This can be fairly easily done with tab completion.

But the real question is why you need to build a specific subset rather than everything? We have put a lot of effort in having the rebuild logic be both minimal and complete.

**Comment by jml1795** [comment]
@jpakkane Agreed, the rebuild logic in meson is excellent and extremely responsive from what I have seen so far. The use case I am trying to implement is building multiple related targets (a library and unit tests or group of related libraries, for example) without having to build everything (in our case, from within a large mono-repo). This would allow building a library or app and the associated tests w/ a single `ninja path/all` invocation.

I wasn't sure if this is something that would be desirable for upstream but I put together an initial implementation (for the ninja backend) inside meson last night in about 8 lines of code, so it was pretty low-impact. I could add it as an opt-in project option so we don't add more targets to existing projects if that is desirable.

This as something akin to package groups without having to explicitly add all the build targets to some list; w/ this change you setup your directory structure to look like the desired package groups and you get a target for the group automatically.

**Comment by jpakkane** [comment]
For that a better solution (and one we have considered but nobody has implemented it yet) is to be able to do something like this:

   meson test --minimal-rebuild testnames

Which would then internally look up everything that the given tests depend on, build only them and then run the tests.

**Comment by jml1795** [comment]
That would be really nice in general - is there another ticket discussing any of the details of its potential implementation? It would solve the build lib + tests example above but doesn't get the ability to rebuild a subset of targets that aren't tied via dependencies directly. For example, in a project structure like
```sh
web/
datamodel/
backend/
guiclient/
utilities/
```
each consisting of many individual targets, we could do a `ninja web/all` without needing to worry about checking for whether guiclient, for example, is up to date (unless you have an explicit dependency on one of its targets from one of web's targets). Maybe meson subprojects ??? would provide a similar capability, but in this case there are build dependencies between the components.

The new phony all targets added to build.ninja would have other preexisting targets as prerequisites; it wouldn't try to implement any new dependency resolution logic.

**Comment by anarazel** [comment]
I'd like to have this feature for reasons not addressed by meson test doing minimal rebuilds: Sometimes when refactoring I just want to rebuild a small part of the code, rather than everything (too noisy, too expensive).

I like @bonzini's suggestion of [#5227](https://github.com/mesonbuild/meson/pull/5227#issuecomment-483585890) using directory/ as the target name.
